### PR TITLE
dispatch.registers to their own file

### LINKED ISSUE #7502: Move pandas dispatch.registers to their own file
In the dask meeting today we talked about how we should be extra careful when making changes to the dispatchers. We are planning on adding some rapids folks to a codeowners file so that they get automatically  pinged when changes happen in the dispatch functions. I started looking into this just now and realized that we register dispatchers for pandas in several different places. I am proposing bringing those all under one file so that we can more easily notice when we make changes to these functions (and it makes the codeowners bit easier :) ).

I will open  PR shortly with the proposed change, but I wanted to throw this idea up in the meantime to see what people think. 

@kkraus14 @jakirkham @rjzamora @jrbourbeau

### PR DESCRIPTION
**Pull request by jsignell**
- [x] Closes #7502
- [x] Tests added / passed
- [x] Passes `black dask` / `flake8 dask` / `isort dask`

These could also go into backends.py if people prefer. I have no strong opinion.

### CODE REVIEW

File: .github/CODEOWNERS
Comment (kkraus14) [inline .github/CODEOWNERS:2]: Ideally we'd target the dispatch methods as opposed to the pandas implementations of the dispatch methods. i.e. https://github.com/dask/dask/blob/main/dask/dataframe/methods.py#L392-L439

Comment (jsignell) [reply .github/CODEOWNERS:2]: Yeah my thinking was that anyone touching the dispatch methods would also change the pandas dispatch methods since that's the way that they'd see any effect. Moving the pandas dispatch registers seemed like a more natural place to separate the code.

Comment (jrbourbeau) [inline .github/CODEOWNERS:2]: I'd prefer to make this a team instead of specific individuals. It will allow us to more easily add/remove folks

Comment (jsignell) [reply .github/CODEOWNERS:2]: Yeah that works for me. I am torn between using dask/gpu,  dask/dataframe, or making a new one that is the intersection.

Comment (jrbourbeau) [reply .github/CODEOWNERS:2]: I think moving both the dispatch objects themselves, as well as the pandas-specific implementations, into a `dask/dataframe/dispatch.py` module seems reasonable from an organizational perspective as it allows us to isolate "these are the things other DataFrame libraries can customize" to a single location. What do you think @jsignell?

Comment (jsignell) [reply .github/CODEOWNERS:2]: Yeah that makes sense to me. Do the top-level functions (like `dd.concat`) then also live in that module?

Comment (jsignell) [reply .github/CODEOWNERS:2]: I think probably the answer should be yes because otherwise won't import order matter?

Comment (jrbourbeau) [reply .github/CODEOWNERS:2]: Also, from [GitHub's docs on code owners](https://docs.github.com/en/github/creating-cloning-and-archiving-repositories/about-code-owners), it looks like code owners (whether they are individuals or teams) must have write permissions to the repository. In order to keep notifications and permissions unrelated, I'll propose we create a `dataframe-dispatch` team (or some other name) that we can manually ping, but hold off on creating a `.github/CODEOWNERS ` file for the time being. This is less ideal than an automated approach, but I suspect will still work in practice

Comment (jrbourbeau) [reply .github/CODEOWNERS:2]: I could be missing something, but I'm not sure why top-level functions like `dd.concat` would be impacted. IIUC we should be able to move all the `Dispatch` objects and their registrations to a separate module and things should work as they do today.

Comment (jsignell) [reply .github/CODEOWNERS:2]: I guess I thought Keith was asking that methods like `concat` also be moved, so that they can more easily be watched. 

As a side-note, I think we need to do a separate module for the Dispatch objects and the panda registrations to avoid circular dependencies.

Comment (jakirkham) [reply .github/CODEOWNERS:2]: Yep agree with this being a team as well

Comment (jrbourbeau) [reply .github/CODEOWNERS:2]: Hmm that's a bummer re: circular dependencies. I'm assuming this is because `concat_pandas` calls `concat` internally so we end up with circular imports. Would something like moving the `concat_dispatch` import to the bottom of `dask/dataframe/methods.py` help?

Comment (jsignell) [reply .github/CODEOWNERS]: I might be wrong. Let me double check.

Comment (jsignell) [reply .github/CODEOWNERS]: https://github.com/dask/dask/pull/7506

**Comment by kkraus14** [comment]
Thanks @jsignell this looks great generally. This is a hugely breaking change that will likely break anyone using DataFrame dispatches outside of Dask (not sure if there's anyone other than dask-cudf).

I'm going to kickoff some conversations internally on our side, but if possible it would be ideal to hold off on merging this for a bit to give us a chance to get our ducks in a row.

**Comment by jsignell** [comment]
> Thanks @jsignell this looks great generally. This is a hugely breaking change that will likely break anyone using DataFrame dispatches outside of Dask (not sure if there's anyone other than dask-cudf).

We should discuss if this is something that is worth doing. To my mind, it's a good clean-up, but if we are committed to breaking people, we should make sure that it's change that we are really happy with.

> I'm going to kickoff some conversations internally on our side, but if possible it would be ideal to hold off on merging this for a bit to give us a chance to get our ducks in a row.

Absolutely

**Comment by kkraus14** [comment]
> We should discuss if this is something that is worth doing. To my mind, it's a good clean-up, but if we are committed to breaking people, we should make sure that it's change that we are really happy with.

I'm a big +1 to this change and it's much appreciated.

Regardless, after the meeting this morning and this RAPIDS release, we've decided as a team that we're going to start constraining Dask to its latest release version at the time of our release instead of letting it float. Breaking changes are going in too frequently and Dask releases go out in between our releases too frequently, where repeatedly fixing our previous release is too expensive for our team to do at this time.

File: dask/dataframe/dispatch.py
Comment (jsignell) [inline dask/dataframe/dispatch.py:77]: It'd be good to get a version number on this that we can put in the _compat file.

**Comment by jsignell** [comment]
This PR would break a lot less if the functions that are being registered were private. Is that already the implication? If it is not then I would be happy to open a smaller PR that prefixes them all with an underscore.

**Comment by jsignell** [comment]
Ok I looked at dask_cudf and tried to preserve the imports that are used there. Would someone be able to run the cudf tests on this PR and let me know if there are any failures?

**Comment by jsignell** [comment]
@kkraus14 and @quasiben can you test this with cudf and let me know if there are failures.

**Comment by jsignell** [comment]
Maybe now that GTC is over :)

**Comment by quasiben** [comment]
@charlesbluca would you have time to run these test and report back ?

**Comment by charlesbluca** [comment]
Ran through the dask-cudf tests, here are the results - lots of failures, but not sure if that's because of my env:

```bash
$ conda list | grep -e "dask\|cudf"
cudf                      0.20.0a210419   cuda_11.2_py38_g867d6eecca_166    rapidsai-nightly
dask                      2.9.0+733.g881a7474           dev_0    <develop>
dask-cudf                 0.20.0a210419   py38_g867d6eecca_166    rapidsai-nightly
libcudf                   0.20.0a210419   cuda11.2_g808262f49c_167    rapidsai-nightly
```

Let me know if I should change anything and rerun the tests.


[test_results.txt](https://github.com/dask/dask/files/6337858/test_results.txt)

**Comment by quasiben** [comment]
here is an example of an error:

```
________________________ test_categorical_basic[data0] _________________________

data = ['a', 'a', 'b', 'c', 'a']
Categories (3, object): ['a', 'b', 'c']

    @pytest.mark.parametrize("data", [data_cat_1()])
    def test_categorical_basic(data):
        cat = data.copy()
        pdsr = pd.Series(cat)
        sr = Series(cat)
        dsr = dgd.from_cudf(sr, npartitions=2)
        result = dsr.compute()
        np.testing.assert_array_equal(cat.codes, result.to_array())

        assert dsr.dtype.to_pandas() == pdsr.dtype
        # Test attributes
        assert pdsr.cat.ordered == dsr.cat.ordered

        assert_eq(pdsr.cat.categories, dsr.cat.categories)

        np.testing.assert_array_equal(pdsr.cat.codes.values, result.to_array())

        string = str(result)
        expect_str = """
    0 a
    1 a
    2 b
    3 c
    4 a
    """
        assert all(x == y for x, y in zip(string.split(), expect_str.split()))

        df = DataFrame()
        df["a"] = ["xyz", "abc", "def"] * 10

        pdf = df.to_pandas()
        cddf = dgd.from_cudf(df, 1)
>       cddf["b"] = cddf["a"].astype("category")

test_accessor.py:142:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../../../../dask-dispatch-test/dask/dataframe/core.py:3885: in __setitem__
    df = self.assign(**{key: value})
../../../../../dask-dispatch-test/dask/dataframe/core.py:4196: in assign
    df2 = self._meta_nonempty.assign(**_extract_meta(kwargs, nonempty=True))
../../../../../dask-dispatch-test/dask/dataframe/core.py:5496: in _extract_meta
    res[k] = _extract_meta(x[k], nonempty)
../../../../../dask-dispatch-test/dask/dataframe/core.py:5488: in _extract_meta
    return x._meta_nonempty if nonempty else x._meta
../../../../../dask-dispatch-test/dask/dataframe/core.py:375: in _meta_nonempty
    return meta_nonempty(self._meta)
../../../../../dask-dispatch-test/dask/utils.py:512: in __call__
    return meth(arg, *args, **kwargs)
../backends.py:96: in _nonempty_series
    data = _get_non_empty_data(s)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

s = Series([], Name: a, dtype: category
Categories (0, dtype: category)

    def _get_non_empty_data(s):
        if isinstance(s._column, cudf.core.column.CategoricalColumn):
            categories = (
                s._column.categories
                if len(s._column.categories)
                else [UNKNOWN_CATEGORIES]
            )
>           codes = column.full(size=2, fill_value=0, dtype="int32")
E           NameError: name 'column' is not defined
```

@galipremsagar @kkraus14 do you have an idea of what's going on here?

**Comment by galipremsagar** [comment]
> >           codes = column.full(size=2, fill_value=0, dtype="int32")
> E           NameError: name 'column' is not defined
> ```
> 
> @galipremsagar @kkraus14 do you have an idea of what's going on here?

This looks like a bug that has to be fixed in `dask-cudf`, `column` -> `cudf.core.column`

**Comment by galipremsagar** [comment]
> This is a bug that has to be fixed in `dask-cudf`, `column` -> `cudf.core.column`

@charlesbluca can you confirm if making this change unblocks you?

**Comment by charlesbluca** [comment]
Yes, that resolves all failures - thanks @galipremsagar! Should I open a PR with cuDF for this?

**Comment by jsignell** [comment]
Thanks @charlesbluca, @quasiben, and @galipremsagar! So seems like this is safe from the dask-cudf perspective?

**Comment by kkraus14** [comment]
> Thanks @charlesbluca, @quasiben, and @galipremsagar! So seems like this is safe from the dask-cudf perspective?

Yes. Generally we are now constraining dask / distributed versions at release time to the latest released patch version so we should be more protected against breaking changes being introduced.

**Comment by jsignell** [comment]
Ok @jrbourbeau should we merge this and #7505?

**Comment by galipremsagar** [comment]
Can we get this PR to be merged? This will help unblock other PRs around having major changes in dispatch: https://github.com/dask/dask/pull/7586

cc: @jakirkham @quasiben @jrbourbeau

**Comment by jakirkham** [comment]
+1 from me

@jrbourbeau do you have any thoughts? Should we merge this and the array PR ( https://github.com/dask/dask/pull/7505 )?

**Comment by jsignell** [comment]
James is fixing up #7505 to match the pattern in this PR and then we'll merge both.

File: dask/dataframe/dispatch.py
Comment (jrbourbeau) [inline dask/dataframe/dispatch.py:4]: Nitpick: Should this be `backends.py`?

Comment (jsignell) [reply dask/dataframe/dispatch.py:4]: I think it this is right. There are some `Dispatch` objects initialized in there as well. I thought they were different in some way and should stay in extension.py, but I can't remember why I thought that. Do you think they should move into this file?

Comment (jrbourbeau) [reply dask/dataframe/dispatch.py:4]: Ah, I see. Since those are specifically for registering extension array, let's keep those where they are

**Review by jrbourbeau — APPROVED** [review]
Thanks @jsignell!

**Comment by chrisroat** [comment]
I think this may have broken the use of make_meta.  (I also see @GenevieveBuckley noted this).

```
>>> dd.utils.make_meta([('a', 'i8'), ('b', 'O')])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/croat/.local/share/virtualenvs/starmap-T47byR32/lib/python3.8/site-packages/dask/utils.py", line 511, in __call__
    meth = self.dispatch(type(arg))
  File "/home/croat/.local/share/virtualenvs/starmap-T47byR32/lib/python3.8/site-packages/dask/utils.py", line 505, in dispatch
    raise TypeError("No dispatch for {0}".format(cls))
TypeError: No dispatch for <class 'list'>
```

**Comment by galipremsagar** [comment]
> I think this may have broken the use of make_meta. (I also see @GenevieveBuckley noted this).
> 
> ```
> >>> dd.utils.make_meta([('a', 'i8'), ('b', 'O')])
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
>   File "/home/croat/.local/share/virtualenvs/starmap-T47byR32/lib/python3.8/site-packages/dask/utils.py", line 511, in __call__
>     meth = self.dispatch(type(arg))
>   File "/home/croat/.local/share/virtualenvs/starmap-T47byR32/lib/python3.8/site-packages/dask/utils.py", line 505, in dispatch
>     raise TypeError("No dispatch for {0}".format(cls))
> TypeError: No dispatch for <class 'list'>
> ```

In https://github.com/dask/dask/pull/7586, there was a new function introduced that has to be invoked to make meta i.e., 
`make_meta_util`: `dd.utils.make_meta_util` and not `make_meta` directly.

**Comment by chrisroat** [comment]
@galipremsagar `make_meta_util` calls `make_meta` for things like a list, so it fails in the same way:

```
$ python
Python 3.8.5 (default, Jan 27 2021, 15:41:15) 
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import dask
>>> dask.__version__
'2021.05.1'
>>> import dask.dataframe as dd
>>> dd.utils.make_meta_util([('a', 'i8'), ('b', 'O')])
Traceback (most recent call last):
  File "/home/croat/make_meta/venv/lib/python3.8/site-packages/dask/dataframe/dispatch.py", line 100, in make_meta_util
    return make_meta(x, index=index)
  File "/home/croat/make_meta/venv/lib/python3.8/site-packages/dask/utils.py", line 511, in __call__
    meth = self.dispatch(type(arg))
  File "/home/croat/make_meta/venv/lib/python3.8/site-packages/dask/utils.py", line 505, in dispatch
    raise TypeError("No dispatch for {0}".format(cls))
TypeError: No dispatch for <class 'list'>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/croat/make_meta/venv/lib/python3.8/site-packages/dask/dataframe/dispatch.py", line 106, in make_meta_util
    func = make_meta_obj.dispatch(type(x))
  File "/home/croat/make_meta/venv/lib/python3.8/site-packages/dask/utils.py", line 505, in dispatch
    raise TypeError("No dispatch for {0}".format(cls))
TypeError: No dispatch for <class 'list'>
>>> 
```

**Comment by jakirkham** [comment]
@chrisroat could you please file this in an issue? This discussion is a bit hard to track in a merged PR, which seems unlikely to be related to the issue at hand

**Comment by chrisroat** [comment]
https://github.com/dask/dask/issues/7731

**Pull request merged**
